emu/src/arm7tdmi.rs

```diff
@@ -460,10 +490,57 @@ impl Arm7tdmi {
         self.cpsr.set_zero_flag(value == 0);
     }
 
+    fn sub_inner_op(first_op: u32, second_op: u32) -> ArithmeticOpResult {
+        let result = first_op.wrapping_sub(second_op);
+
+        let sign_op1: bool = first_op.get_bit(31);
+        let sign_op2: bool = second_op.get_bit(31);
+        let sign_r: bool = result.get_bit(31);
+
+        let different_sign = sign_op1 != sign_op2;
+
+        ArithmeticOpResult {
+            result,
+            carry: first_op < second_op,
+            overflow: different_sign && sign_op2 == sign_r,
+            sign: result.get_bit(31),
+            zero: result == 0,
+        }
+    }
+
+    fn cmn(&mut self, rn: u32, op2: u32) {
+        let add_result = Self::add_inner_op(rn, op2);
+
+        self.cpsr.set_overflow_flag(add_result.overflow);
+        self.cpsr.set_carry_flag(add_result.carry);
+        self.cpsr.set_zero_flag(add_result.zero);
+        self.cpsr.set_sign_flag(add_result.sign);
+    }
+
     fn cmp(&mut self, rn: u32, op2: u32) {
-        let value = rn - op2;
-        self.cpsr.set_sign_flag(value.is_bit_on(31));
-        self.cpsr.set_zero_flag(value == 0);
+        let sub_result = Self::sub_inner_op(rn, op2);
+
+        self.cpsr.set_overflow_flag(sub_result.overflow);
+        self.cpsr.set_carry_flag(sub_result.carry);
+        self.cpsr.set_zero_flag(sub_result.zero);
+        self.cpsr.set_sign_flag(sub_result.sign);
+    }
+
+    fn sub(&mut self, rd: usize, rn: u32, op2: u32, s: bool) {
+        let sub_result = Self::sub_inner_op(rn, op2);
+
+        self.registers.set_register_at(rd, sub_result.result);
+
+        if s {
+            self.cpsr.set_overflow_flag(sub_result.overflow);
+            self.cpsr.set_carry_flag(sub_result.carry);
+            self.cpsr.set_zero_flag(sub_result.zero);
+            self.cpsr.set_sign_flag(sub_result.sign);
+        }
```

Comment:
Would love to see `sub_result` passed directly to `cpsr` in some way, because seems an over engineering with this extra struct. Using a separated fn for creating a struct and other bunch of line for assigning... Don't sure we have a better design

Reply:
The separate fn is in my opinion needed since the add/sub logic is used in many ops. What would you return from this fn? At first, I thought about a tuple but a struct seemed more clear since it has named properties.

We could add a fn in `Cpsr` that accepts an `ArithmeticOpResult` and sets the flags?

I don't know what would be the best solution in this case

Reply:
We can go for something like `self.cpsr.apply_flags(flags_result)`

Reply:
Done, thanks!